### ISSUE TITLE
Unable to add numeric types that are not simply 'number'

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#17069]
I'm trying to rewrite the signature of `sum` from `core.ts` in our compiler (see #16823), but I'm getting this error:

```ts
export function sum<T extends Record<K, number>, K extends string>(array: T[], prop: K): number {
    let result = 0;
    for (const v of array) {
        result += v[prop];
     // ~~~~~~~~~~~~~~~~~
     // Operator '+=' cannot be applied to types 'number' and 'T[K]'.
    }
    return result;
}
```

Note that I can't declare `sum` as

```ts
export function sum<K extends string>(array: Record<K, number>[], prop: K): number
```

because then TypeScript would infer *all* of the properties in each element for `K`. So I do need another type argument `T` for the indirection.

For a simpler repro:

```ts
function foo<T extends number>(x: T) {
    return 0 + x;
    //     ~~~~~
    // Operator '+' cannot be applied to types '0' and 'T'.

}
```

Related is #15645.

**Comment by andy-ms** [comment]
Related:  #15501

**Comment by sandersn** [comment]
Fix is up at #17870

**Comment by sandersn** [comment]
Note that #17404 fixed the simpler repro; #17870 fixes the original issue with a type parameter indexed by another type parameter.

**Comment by sandersn** [comment]
The previous PR got merged by mistake and is now reverted; #17912 is the PR with the fix now.
